osquery/events/linux/containereventpublisher.cpp

```diff
@@ -0,0 +1,560 @@
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include <regex>
+#include <unordered_map>
+
+#include <boost/asio.hpp>
+
+#include <grpcpp/grpcpp.h>
+
+#include <google/protobuf/util/time_util.h>
+
+#include <osquery/core/flags.h>
+#include <osquery/events/containerd/container.grpc.pb.h>
+#include <osquery/events/containerd/content.grpc.pb.h>
+#include <osquery/events/containerd/image.grpc.pb.h>
+#include <osquery/events/containerd/namespace.grpc.pb.h>
+#include <osquery/events/containerd/runc.grpc.pb.h>
+#include <osquery/events/containerd/snapshot.grpc.pb.h>
+#include <osquery/events/containerd/task.grpc.pb.h>
+#include <osquery/events/eventsubscriber.h>
+#include <osquery/events/linux/containereventpublisher.h>
+#include <osquery/logger/logger.h>
+#include <osquery/registry/registry_factory.h>
+#include <osquery/utils/json/json.h>
+
+#include <boost/algorithm/string/classification.hpp>
+#include <boost/algorithm/string/split.hpp>
+
+namespace local = boost::asio::local;
+
+namespace osquery {
+
+const std::string kContainerEventTypePrefix = "containerd.events.Container";
+const std::string kContentEventTypePrefix = "containerd.events.Content";
+const std::string kImageEventTypePrefix = "containerd.events.Image";
+const std::string kNamespaceEventTypePrefix = "containerd.events.Namespace";
+const std::string kSnapshotEventTypePrefix = "containerd.events.Snapshot";
+const std::string kTaskEventTypePrefix = "containerd.events.Task";
+
+REGISTER(ContainerEventPublisher, "event_publisher", "containerevent");
+
+FLAG(string,
+     containerd_socket,
+     "/run/containerd/containerd.sock",
+     "Docker UNIX domain socket path");
+
+FLAG(string,
+     containerd_events_namespaces,
+     "",
+     "Comma separated list of namespaces to receive events from, empty means "
+     "all");
+
+FLAG(bool,
+     enable_containerd_events_publisher,
+     false,
+     "Enable the containerd event publisher");
+
+namespace {
+bool IsPublisherEnabled() noexcept {
+  return FLAGS_enable_containerd_events_publisher;
+}
+
+template <typename T, typename... ConstructorArgs>
+auto createContainerEvent(
+    const containerd::services::events::v1::Envelope& event,
+    ConstructorArgs&&... args) {
+  return std::make_unique<T>(
+      google::protobuf::util::TimeUtil::ToString(event.timestamp()),
+      event.namespace_(),
+      event.topic(),
+      std::forward<ConstructorArgs>(args)...);
+}
+
+std::unordered_map<std::string, std::string> labelsProtobufMapToUnorderedMap(
+    const google::protobuf::Map<std::string, std::string>& protobuf_map) {
+  std::unordered_map<std::string, std::string> labels;
+
+  for (const auto& label_pair : protobuf_map) {
+    labels.emplace(label_pair.first, label_pair.second);
+  }
+
+  return labels;
+}
+
+} // namespace
+
+Status ContainerEventPublisher::setUp() {
+  if (!IsPublisherEnabled()) {
+    return Status::failure("Publisher disabled via configuration");
+  }
+
+  grpc_init();
```

Comment:
i guess, `grpc_init()` doesn't return any status code..? how likely is it that `grpc_init()` fails?

Reply:
The definition is `void grpc_init(void)` so there's no real way of telling if something failed. The function calls out a lot but they all seem to be allocation and structure setup. Technically a plugin could do something more complicated but we shouldn't be using any.
We could catch bad_alloc but I don't think osquery deals with allocation failures properly anywhere else

Reply:
It seems like other places do catch bad_alloc so I added that